### DISCUSSION TITLE
Warnings vs errors vs extensions vs flags

### DISCUSSION THREAD

**Discussion by goldfirere** [ghc-proposals/ghc-proposals#502] [Ideas]
GHC today uses several different mechanisms to achieve the same end goal: eliminating code that is somehow suboptimal, but can be accepted anyway. This discussion is meant to be a place to consider whether we should choose one such mechanism going forward, and whether we should try to clean up the current mess.

### Flexible contexts

If I specify `-XHaskell2010` and write a type signature `f :: C Int a => ...`, GHC will reject because the context has a funny shape. Enabling `-XFlexibleContexts` allows GHC to accept the program. Contexts like `C Int a` can sometimes cause unpredictable type inference, and so a user might plausibly want to avoid them. Enabling `-XFlexibleContexts` causes no change to GHC's behavior other than allowing it to accept more programs (and giving them runtime semantics).

### Deferred type errors

If I write an expression `not 'x'`, GHC will reject because there is a type error. Enabling `-fdefer-type-errors` allows GHC to accept the program with a warning. Enabling `-Wno-deferred-type-errors` disables the warning. Expressions like `not 'x'` will error at runtime, and so a user might plausibly want to avoid them. Enabling `-fdefer-type-errors` causes no change to GHC's behavior other than allowing it to accept more programs (and giving them runtime semantics).

### Partial type signatures

If I write a type signature `f :: Int -> _`, GHC will reject because there is a hole in the type. Enabling `-XPartialTypeSignatures` allows GHC to accept the program with a warning. Enabling `-Wno-partial-type-signatures` disables the warning. Type signatures like `Int -> _` might be desired only during development, not in production, and so a user might plausibly want to avoid them. Enabling `-XPartialTypeSignatures` causes no change to GHC's behavior other than allowing it to accept more programs (and giving them runtime semantics).

### Incomplete pattern matches

If I specify `-Werror=incomplete-patterns` and write an expression `\x -> (case x of Nothing -> 5)`, GHC will reject because the pattern-match is incomplete. Enabling `-Wno-incomplete-patterns` allows GHC to accept the program. Expressions like `\x -> (case x of Nothing -> 5)` can error at runtime, and so a user might plausibly want to avoid them. Enabling `-Wno-incomplete-patterns` causes no change to GHC's behavior other than allowing it to accept more programs (and giving them runtime semantics).

### Implicit `forall`

(With accepted proposal #285) If I specify `-XNoImplicitForAll` and write a top-level type signature `f :: a -> a`, GHC will reject because `a` is out of scope. Enabling `-XImplicitForAll` allows GHC to accept the program. Type signatures like `f :: a -> a` have a variable occurrence with no binding site, and so a user might plausibly want to avoid them. Enabling `-XImplicitForAll` causes no change to GHC's behavior other than allowing it to accept more programs (and giving them runtime semantics).

## Analysis

We see here that GHC's way of discouraging some code patterns varies wildly. Sometimes the pattern is allowed by default but can be disabled with an extension. Sometimes the pattern is disallowed by default but can be enabled with an extension. Sometimes a warning appears even after the pattern is enabled, sometimes not. This status quo requires users to be aware of a variety of mechanisms for getting their code to compile, and it complicates the implementation, which must deal with all of these scenarios.

## A modest proposal

I propose that, going forward, we handle all such occurrences via the warnings mechanism, but with a new twist: some warnings will be `Werror` by default. (No warnings are `Werror` by default currently.) That is, a vanilla invocation of GHC will have `-Werror=deferred-type-errors -Werror=partial-type-signatures` and, in `-XHaskell2010` mode, `-Werror=flexible-contexts` (among others). Proposal #285 would then change to introduce warning flags `implicit-for-all` and `pattern-signature-binds`, included in no warning bucket, but with the possibility of being enabled by users who want them.

This new design would be uniform and easier to implement. Tool support would likely be easier, too. Taken to its logical conclusion, it would mean that all extensions actually change the meaning of existing code, because any extension that doesn't would be implemented via the warnings mechanism.

-----------------------------

What do we think? I'm not actually suggesting we start a program of changing all existing extensions in this way. More that we should design future features around this principle. Maybe, later, if we like how it's all going, we can think about changing old extensions. Most immediately, this would change #285 (and its reincarnation in #448) to use the warnings mechanism over the extensions mechanism, which seems like a nice simplification (and is the direct motivation for this Discussion). But users have also been surprised that diagnostic-free partial type signatures require both `-XPartialTypeSignatures` *and* `-Wno-partial-type-signatures`. In the GHC implementation, the fact that both `-Werror` and `-fdefer-type-errors` change errors to warnings is a small spot of awkwardness that would be cleaned up.

**Comment by Ericson2314** [comment]
I am not so sure. I think language extensions and warnings have different normative connotations. It is useful to track what is some ad-hoc analysis, and what is a candidate for inclusion in a future Haskell report.

Perhaps a good first step would be to get the current Haskell Report and put in a repo so we can have a "living" Haskell Report. If we have a proposal process for changing the Haskell report, we can work "backwards" and decide to what extent language extensions in GHC should be "ready to propose as report changes".

That might sounds a bit silly since for all practical purpose "Haskell is dead, long live GHC". Still, I would point out that our Kin in Rust absolutely does distinguish (the disabling of) unstable features from other lints.

-----

If this sounds funny from me coming from https://github.com/ghc-proposals/ghc-proposals/pull/351, which was shot down in part because the "too many knobs" issue. Well, I am confident that I now have a way to formalize it that *does* make `-XNoIncomplete` a "real" language extension decoupled form the completely informal pattern match completeness checker. (The lawyering is predicated on absurd alternatives -- `-XNoIncomplete` simply means you must handle every pattern according to H98-style simple checking of patterns of each alternative. What is *not* part of `-XNoIncomplete` is when GHC will accept an absurd alternative -- the absurd alternative *must* be explicitly written out for any impossible cases rather than empty-case style omissions. but one could always do a `_ -> error "whoops"` too the satisfy `-XNoIncomplete` without fancy stuff too.)

I am quite proud of this crafty little bit of indirection. If we did what you were proposing above, perhaps we would already have `-Wincompleteness`, but I would have never been *pushed* by the proposal process to come up with this separation of concerns (simple completeness checking fancy absurd pattern checking) that I think is better regardless of where the bar for a language extension like `-XNoIncomplete` is set. I therefore appreciate the current process, whatever its faults, for pushing me to improve my design.

**Comment by Ericson2314** [comment]
On a separate note, as @hsyl20 @doyougnu and I will be releasing a document soon describing, `OPTIONS_GHC` is a terrible layer violation with real pragmatic problems too (huge source of bugs, we completely abrogated a responsible to spec out what is allowed in `OPTIONS_GHC`, etc.). Until we have a `{-# WARNING .... #-}` pragma or similar, I cannot endorse any plan changing language extensions to warnings, as that will just lead to more `OPTIONS_GHC` sinning.

**Comment by treeowl** [comment]
Some of these things have become local pragmas. Others (typed holes and ambiguous types) seem destined to do the same. That is one great way to resolve tension. `AllowAmbiguousTypes` is also likely to be (mostly?) obviated by visible type arguments or whatever those are called. Some extensions have important interactions (`MonoLocalBinds` with `TypeFamilies` and `GADTs`). I know you're talking about future and not past, but there's no reason to think new extensions will develop much differently.

**Comment by nomeata** [comment]
The perception that all of these things are best just warnings that could be errors is a bit post-hoc, taking the “maximal GHC Haskell” (all extensions enabled) as the starting point and noticing that really, turning of an extension is similar to introducing a warning for that feature, and making that warning fatal.

But I’m not sure that it does justice to the extension menchanism in general, and only seems to work well if there is a single “maximal GHC Haskell”, and not for example forking or mutually exclusive extensions, for example.

So I assume you are not proposing to handle all, or even most, future language extensions via “add the new feature, define a warning for using the new feature, add it to `-Werror`”, right? But which one do you want to propose to be implemented that way?

Also, what will happen to `-Wall` or `-Weverything`?

It seems a bit odd to have something called a “warning” that isn’t (generally) considered bad or discouraged, but merely “new” and “hasn’t been around forever, so can’t be on by default”. I guess all examples above (including `pattern-signature-binds`) could be argued fall under “bad or discouraged, if only because we envision better ways”. But would we have something like `-Werror=explicit-let-rec` to guard a potential `let rec` syntactic extension (together with an optional `-Wwarn=implicit-recursion` to complain about implicit recursion)?
